log/record_test.go

```diff
@@ -103,3 +103,101 @@ func TestRecordAttributes(t *testing.T) {
 		}
 	})
 }
+
+func BenchmarkRecord(b *testing.B) {
+	var (
+		tStamp time.Time
+		sev    log.Severity
+		text   string
+		body   log.Value
+		attr   log.KeyValue
+		n      int
+	)
+
+	b.Run("Timestamp", func(b *testing.B) {
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			var r log.Record
+			r.SetTimestamp(y2k)
+			tStamp = r.Timestamp()
+		}
+	})
+
+	b.Run("ObservedTimestamp", func(b *testing.B) {
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			var r log.Record
+			r.SetObservedTimestamp(y2k)
+			tStamp = r.ObservedTimestamp()
+		}
+	})
+
+	b.Run("Severity", func(b *testing.B) {
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			var r log.Record
+			r.SetSeverity(log.SeverityDebug)
+			sev = r.Severity()
+		}
+	})
+
+	b.Run("SeverityText", func(b *testing.B) {
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			var r log.Record
+			r.SetSeverityText("text")
+			text = r.SeverityText()
+		}
+	})
+
+	bodyVal := log.BoolValue(true)
+	b.Run("Body", func(b *testing.B) {
+		b.ReportAllocs()
+		for n := 0; n < b.N; n++ {
+			var r log.Record
+			r.SetBody(bodyVal)
+			body = r.Body()
+		}
+	})
+
+	attrs10 := []log.KeyValue{
+		log.Bool("b1", true),
+		log.Int("i1", 324),
+		log.Float64("f1", -230.213),
+		log.String("s1", "value1"),
+		log.Map("m1", log.Slice("slice1", log.BoolValue(true))),
+		log.Bool("b2", false),
+		log.Int("i2", 39847),
+		log.Float64("f2", 0.382964329),
+		log.String("s2", "value2"),
+		log.Map("m2", log.Slice("slice2", log.BoolValue(false))),
+	}
```

Comment:
The attributes should be inlined in benchmark code.  The benchmark should represent regular API usage. The benchmark will then show show that the code is written in a way that the Go escape analysis makes sure that it will not cause any allocation.

For reference this is also how the slog benchmarks are written. E.g. 
- https://github.com/golang/go/blob/master/src/log/slog/internal/benchmarks/benchmarks_test.go
- https://github.com/golang/go/blob/master/src/log/slog/record_test.go#L147

When working on the prototype there would be significant differences if the attributes would be pre-allocated: https://github.com/open-telemetry/opentelemetry-go/blob/147762ffb5c4394417b3f973274cc570723db79b/log/internal/bench_test.go. E.g.g this benchmark should show why we have not ended up with:

```go
type Record struct {
	Timestamp         time.Time
	ObservedTimestamp time.Time
	Severity          Severity
	SeverityText      string
	Body              Value
	Attributes        []KeyValue
}
```

Reply:
Haven't we already had [this discussion](https://github.com/open-telemetry/opentelemetry-go/pull/3971#discussion_r1163272609)?